Make LittleFS Flash driver integration a configuration option

mikk-leini-krakul opened the pull request:
Not all systems use LittleFS on Flash memory. They may use SD card or eMMC as physical media, hence adding Flash driver is unwanted payload. Some SoCs do not have Flash driver in Zehpyr, that was actually the reason to create this PR.
Added configuration option to enable/disable LittleFS Flash memory interface, just like there's configuration option for disk access layer interface. For backward compatibility reasons, have Flash driver interface enabled by default.

mikk-leini-krakul committed:
Make LittleFS flash storage interface a configuration option.

Add Kconfig parameter FS_LITTLEFS_FLASH_DEV.

Laczen commented:
@mikk-leini-krakul, can both options be enabled at the same time ?

mikk-leini-krakul commented:
> @mikk-leini-krakul, can both options be enabled at the same time ?

Yes

de-nordic commented on subsys/fs/littlefs_fs.c:
```diff
@@ -803,19 +830,30 @@ static int littlefs_mount(struct fs_mount_t *mountp)
-		lcp->read = lfs_api_read;
-		lcp->prog = lfs_api_prog;
-		lcp->erase = lfs_api_erase;
+		lcp->read = lfs_api_read_flash;
+		lcp->prog = lfs_api_prog_flash;
+		lcp->erase = lfs_api_erase_flash;
```
Why change names? They do not collide with anything.

mikk-leini-krakul replied:
To have a naming pattern and future proofing the next memory access interface. Function without suffix sounds like an entry point for all other interfaces, but it isn't.

de-nordic replied:
But the commit does not add one, so no need to rename it for now.

Laczen commented:
@mikk-leini-krakul, I will remove myself as a reviewer because I am somewhat biased as I proposed the exact same thing in #41972. As this was rejected for silly reasons I cannot spend any time on your proposal.

mikk-leini-krakul commented:
> 

I have always thought that when several independent engineers come to same solution then it is a worthy solution. I don't see a reason why you shouldn't contribute.

mikk-leini-krakul commented on subsys/fs/littlefs_fs.c:
```diff
@@ -708,16 +710,28 @@ static int littlefs_mount(struct fs_mount_t *mountp)
 				LOG_ERR("Unable to get sector size");
 				goto out;
 			}
+#else
+			LOG_ERR("LittleFS disc access not enabled");
+			ret = -ENOSYS;
+			goto out;
+#endif
```
You're right, there are excessive checks.

mikk-leini-krakul committed:
Implement LittleFS flash storage / disk access interface selection.

Take the FS_LITTLEFS_FLASH_DEV parameter into use, to compile LittleFS flash storage interface or not.

mikk-leini-krakul commented on subsys/fs/Kconfig.littlefs:
```diff
@@ -105,6 +103,13 @@ config FS_LITTLEFS_HEAP_PER_ALLOC_OVERHEAD_SIZE
+	help
+	  Enable this option to provide support for littlefs on the Flash
+	  memory devices (like for example MCU embedded or (Q)SPI Flash).
+
```
Added. Also restored the dependencies.

mikk-leini-krakul commented:
> Add commit with release note informing on addition of the Kconfig. Work on LittleFS tests and samples to test the flag where applicable.

I rearranged commits so there's clear configuration parameter commit and its implementation commit. 
Will work on tests & samples.

mikk-leini-krakul commented on subsys/fs/littlefs_fs.c:
```diff
@@ -60,8 +64,7 @@ static K_HEAP_DEFINE(file_cache_heap, CONFIG_FS_LITTLEFS_FC_HEAP_SIZE);
 {
-	return IS_ENABLED(CONFIG_FS_LITTLEFS_BLK_DEV) &&
-		mountp->flags & FS_MOUNT_FLAG_USE_DISK_ACCESS;
+	return mountp->flags & FS_MOUNT_FLAG_USE_DISK_ACCESS;
```
How tough is that flag specification?

If I may argue here - automatically switching from flash storage to disk access and vice versa is masking potential fault in user code and it is unwanted behavior on system where both are made available. If third interface is added in the future, the logic gets unnecessarily confusing, but it should be simple and straightforward.

I listed all combinations by the spec, by the old code and new code:
![image](https://user-images.githubusercontent.com/91126161/198218514-d2c0cebd-9f9c-4efb-b59d-c5b37505c34d.png)

6th combination (green cells) changes the behavior. It would be a good time to change the spec and behavior if wanted because later change causes also 3rd combination change.

de-nordic replied:
OK then, we should change how the flag works, this requires stable API change process and changes to FAT FS too.

de-nordic commented:
@mikk-leini-krakul in summary
 1) We can follow with change of meaning for the flag, but need to follow stable API change;
 2) FAT FS needs change to error when flag is not set, and this needs fixing:
 https://github.com/zephyrproject-rtos/zephyr/blob/96698450c49b08b93f71d88f35560a5d820ecd8e/subsys/fs/fat_fs.c#L444-L446
 3) Tests need to be altered for LittleFS and FAT FS;
 4) release notes entry needs to be added.

And don't rename anything that does not require changes at this time, it just adds to diff in PR.
